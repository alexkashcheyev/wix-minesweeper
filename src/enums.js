export const actionType = {
    TOGGLE_MENU: 'TOGGLE_MENU',
    TOGGLE_SUPERMAN: 'TOGGLE_SUPERMAN',
    SET_NEW_GAME_PARAMETER: 'SET_NEW_GAME_PARAMETER',
    START_GAME: 'START_GAME',
    TOGGLE_FLAG: 'TOGGLE_FLAG',
    OPEN_CELL: 'OPEN_CELL',
    MOVE_VIEWPORT: 'MOVE_VIEWPORT',
    SET_MESSAGE: 'SET_MESSAGE',
    SET_FIELD: 'SET_FIELD',
    UPDATE_FLAG_COUNT: 'UPDATE_FLAG_COUNT',
    SET_VIEWPORT_OFFSET: 'SET_VIEWPORT_OFFSET',
    SET_GAME_STAGE: 'SET_GAME_STAGE',
    SET_VIEWPORT_SIZE: 'SET_VIEWPORT_SIZE'
}

export const gameStage = {
    NOT_STARTED: 'NOT_STARTED',
    STARTED: 'STARTED',
    WON: 'WON',
    LOST: 'LOST'
}