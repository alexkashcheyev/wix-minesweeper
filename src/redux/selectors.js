export const selectCurrentGameInfo = (state) => state.currentGame.gameInfo;

export const selectCurrentGame = (state) => state.currentGame;

export const selectField = (state) => state.currentGame.field;