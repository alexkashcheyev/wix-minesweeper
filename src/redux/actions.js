import { actionType } from '../enums';

// request to show or hide the menu
export const toggleMenu = (show) => ({
    type: actionType.TOGGLE_MENU,
    payload: { show }
});

// set superman mode to given value
export const toggleSuperman = (superman) => ({
    type: actionType.TOGGLE_SUPERMAN,
    payload: { superman }
});

// set new game parameter, which is 'width', 'height' or 'mines'
// to given value, if valid
export const setNewGameParameter = (key, value) => ({
    type: actionType.SET_NEW_GAME_PARAMETER,
    payload: { key, value }
});

// request starting game
export const startGame = (gameInfo) => ({
    type: actionType.START_GAME,
    payload: { gameInfo }
})

// set or unset flag on given cell
export const toggleFlag = (x, y) => ({
    type: actionType.TOGGLE_FLAG,
    payload: { x, y }
});

// request opening a cell
export const openCell = (x, y) => ({
    type: actionType.OPEN_CELL,
    payload: { x, y }
})

// request to move the viewport to see different part of the field
export const moveViewport = (dx, dy) => ({
    type: actionType.MOVE_VIEWPORT,
    payload: { dx, dy }
})

// show/hide the alert with given data
export const setMessage = (visible, severity='info', title='', content='') => ({
    type: actionType.SET_MESSAGE,
    payload: { severity, visible, title, content }
})

// set the game field as a reflection
// of user's actions
export const setField = (field) => ({
    type: actionType.SET_FIELD,
    payload: { field }
})

// increase flags count by given number
export const changeSetFlags = (delta) => ({
    type: actionType.UPDATE_FLAG_COUNT,
    payload: { delta }
})

// actually set the viewport offset
export const setViewportOffset = (offset) => ({
    type: actionType.SET_VIEWPORT_OFFSET,
    payload: { offset }
})

// move game to another stage
export const setGameStage = (stage) => ({
    type: actionType.SET_GAME_STAGE,
    payload: { stage }
})