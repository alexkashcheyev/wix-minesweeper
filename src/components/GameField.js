import React from 'react';
import { connect } from 'react-redux';
import { makeStyles } from '@material-ui/core';

import Cell from './Cell'
import { gameStage } from '../enums';
import * as actions from '../redux/actions';
import config from '../appconfig';

const useStyles = makeStyles((theme) => ({
    root: {
        display: 'flex',
        flexDirection: 'row',
        borderWidth: '1px',
        borderStyle: 'solid',
        borderTopColor      : props => props.border.top     ? theme.palette.primary.light : 'transparent',
        borderBottomColor   : props => props.border.bottom  ? theme.palette.primary.light : 'transparent',
        borderLeftColor     : props => props.border.left    ? theme.palette.primary.light : 'transparent',
        borderRightColor    : props => props.border.right   ? theme.palette.primary.light : 'transparent'
    },
    column: {
        display: 'flex',
        flexDirection: 'column',
        justifyContent: 'center'
    }
}));

function buildCell({x, y, key, cell, classes, superman, stage, handleFlag, handleOpen}) {
    return (
        <Cell
            className={classes.cell}
            key={key}
            cell={cell}
            superman={superman}
            size={config.cellSize}
            revealMines={stage === gameStage.LOST}
            disabled={stage === gameStage.LOST || stage === gameStage.WON}
            onFlag={() => handleFlag(x, y)}
            onOpen={() => handleOpen(x, y)}
        />
    )
}

function buildColumns({ field, gameInfo, viewport, classes, superman, stage, handleFlag, handleOpen }) {
    return field
    
        // filter only visible columns
        .filter(
            (column, x) =>
                gameInfo.width <= viewport.width
                || (
                    x >= viewport.offset.x
                    && x < viewport.offset.x + viewport.width
                )
        ).map((column, x) => {

            const cells = column.

                // filter only visible cells
                filter(
                    (cell, y) =>
                        gameInfo.height <= viewport.height
                        || (
                            y >= viewport.offset.y
                            && y < viewport.offset.y + viewport.height
                        )
                ).map((cell, y) => {

                    // x and y here are relative to the viewport,
                    // not to the actual field

                    const fieldX = x + viewport.offset.x;
                    const fieldY = y + viewport.offset.y;

                    // a unique key for reactjs
                    const key = fieldX + '_' + fieldY;
                    
                    return buildCell({ x: fieldX, y: fieldY, key, cell, classes, superman, stage, handleFlag, handleOpen })
                });

            return (
                <div className={classes.column} key={x}>
                    {cells}
                </div>
            );
        })
}

function GameField({ field, dispatch, superman, gameInfo, viewport, stage, border }) {
    
    const classes = useStyles({ border });

    function handleFlag(x, y) {
        dispatch(actions.toggleFlag(x, y));
    }

    function handleOpen(x, y) {
        dispatch(actions.openCell(x, y));
    }

    const columns = buildColumns({ field, gameInfo, viewport, classes, superman, stage, handleFlag, handleOpen });

    return (
        <div className={classes.root} >
            {columns}
        </div>
    )
}


function mapStateToProps(state) {
    return state.currentGame;
}

export default connect(mapStateToProps)(GameField)