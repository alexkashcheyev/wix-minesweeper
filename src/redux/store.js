import { createStore, applyMiddleware } from 'redux';
import { gameStage, actionType } from '../enums';
import createSagaMiddleware from 'redux-saga';
import toggleFlagSaga from '../sagas/toggleFlagSaga';
import generateFieldSaga from '../sagas/generateFieldSaga';
import changeOffsetSaga from '../sagas/changeOffsetSaga';
import openCellSaga from '../sagas/openCellSaga';
import { fork, all } from 'redux-saga/effects';
import config from '../appconfig';

const intialState = {
    ui: {
        showMenu: false,
        message: {
            severity: 'info',
            visible: false,
            title: '',
            content: ''
        }
    },
    currentGame: {
        gameInfo: {
            width: 15,
            height: 15,
            mines: 10
        },
        viewport: {
            width: config.viewportWidth,
            height: config.viewportHeight,
            offset: {
                x: 0,
                y: 0
            },
        },
        stage: gameStage.NOT_STARTED,
        superman: true,
        flagsSet: 0,
        field: []
    },
    newGame: {
        gameInfo: {
            width: 300,
            height: 300,
            mines: 1
        }
    }
}

const rootReducer = (state = intialState, action) => {
    switch (action.type) {

        case actionType.TOGGLE_MENU:
            return {
                ...state,
                ui: {
                    ...state.ui,
                    showMenu: action.payload.show
                }
            }

        case actionType.TOGGLE_SUPERMAN:
            return {
                ...state,
                currentGame: {
                    ...state.currentGame,
                    superman: action.payload.superman
                }
            }

        case actionType.SET_NEW_GAME_PARAMETER:
            return {
                ...state,
                newGame: {
                    ...state.newGame,
                    gameInfo: {
                        ...state.newGame.gameInfo,
                        [action.payload.key]: action.payload.value
                    }
                }
            }

        case actionType.START_GAME:
            return {
                ...state,
                currentGame: {
                    ...state.currentGame,
                    ...action.payload,
                    flagsSet: 0,
                    stage: gameStage.LOADING,
                    viewport: {
                        ...state.currentGame.viewport,
                        offset: {
                            x: 0,
                            y: 0
                        }
                    }
                },
            }

        case actionType.SET_MESSAGE: {

            return {
                ...state,
                ui: {
                    ...state.ui,
                    message: action.payload
                }
            }
            
        }

        case actionType.SET_FIELD: {

            return {
                ...state,
                currentGame: {
                    ...state.currentGame,
                    field: action.payload.field
                }
            }
        }

        case actionType.UPDATE_FLAG_COUNT: {
            
            return {
                ...state,
                currentGame: {
                    ...state.currentGame,
                    flagsSet: state.currentGame.flagsSet + action.payload.delta
                }
            }
        }

        case actionType.SET_VIEWPORT_OFFSET: {

            return {
                ...state,
                currentGame: {
                    ...state.currentGame,
                    viewport: {
                        ...state.currentGame.viewport,
                        offset: action.payload.offset                        
                    }
                }
            }
        }

        case actionType.SET_GAME_STAGE: {
            return {
                ...state,
                currentGame: {
                    ...state.currentGame,
                    stage: action.payload.stage
                }
            }
        }

        default: return state;
    }
}

const sagaMiddleware = createSagaMiddleware();

export const store = createStore(
    rootReducer, 
    applyMiddleware(sagaMiddleware)
);

function* rootSaga () {
    yield all([
        fork(toggleFlagSaga),
        fork(generateFieldSaga),
        fork(changeOffsetSaga),
        fork(openCellSaga)
    ]);
}

sagaMiddleware.run(rootSaga);