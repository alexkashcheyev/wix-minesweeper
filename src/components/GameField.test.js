import React from 'react';
import TestRenderer, { act } from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider }from 'react-redux';

import { gameStage } from '../enums';
import { initialState } from '../redux/store';
import GameField from "./GameField";

describe('GameField component', () => {
    const mockStore = configureStore([]);
    let store;
    let tr;
    let component;

    const wrappedJsx = (store) => (
        <Provider store={store}>
            <GameField border={defaultBorder} />
        </Provider>
    )

    const rerender = () => {
        act(
            () => {
                tr.update(wrappedJsx(store));
            }
        )
        component = tr.toJSON();
    }

    beforeEach(() => {
        store = mockStore(initialState);

        act(() => {
            tr = TestRenderer.create(wrappedJsx(store));
        });

        component = tr.toJSON();
    })    

    it('should create', () => {
        expect(component).toBeTruthy();
    });

    it.each`
        viewportWidth | viewportHeight | expectedWidth | expectedHeight 
        ${3}          | ${4}           | ${3}          | ${4}
        ${5}          | ${4}           | ${5}          | ${4}
        ${7}          | ${4}           | ${5}          | ${4}

        ${3}          | ${5}           | ${3}          | ${5}
        ${5}          | ${5}           | ${5}          | ${5}
        ${7}          | ${5}           | ${5}          | ${5}

        ${3}          | ${8}           | ${3}          | ${5}
        ${5}          | ${8}           | ${5}          | ${5}
        ${7}          | ${8}           | ${5}          | ${5}
    `('should take correct dimensions from viewport and/or field', ({
        viewportWidth, 
        viewportHeight, 
        expectedWidth, 
        expectedHeight
    }) => {
        
        const fieldWidth = 5;
        const fieldHeight = 5;
        // empty field would normally be invalid, but the 
        // component does not care about validity. 

        const field = new Array(fieldWidth)
            .fill(0)
            .map(
                () => new Array(fieldHeight).fill({...defaultCell})
            );
        
        store = mockStore({
            ...initialState,
            currentGame: {
                ...initialState.currentGame,
                field: field,
                gameInfo: {
                    ...initialState.currentGame.gameInfo,
                    width: fieldWidth,
                    height: fieldHeight,
                },
                viewport: {
                    ...initialState.currentGame.viewport,
                    width: viewportWidth,
                    height: viewportHeight,
                }
            }
        })

        rerender();

        expect(component).toMatchSnapshot();
        expect(component.children.length).toEqual(expectedWidth);
        for(let col of component.children) {
            expect(col.children.length).toEqual(expectedHeight);
        }
    });

});

const defaultBorder = {
    borderLeftColor: 'transparent',
    borderTopColor: 'transparent',
    borderRightColor: 'transparent',
    borderBottomColor: 'transparent',
}

const defaultCell = {
    isOpened: false,
    hasMine: false,
    isFlagged: false,
    minesAround: 0,
}