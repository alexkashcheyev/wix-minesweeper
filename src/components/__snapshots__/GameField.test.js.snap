// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameField component should take correct dimensions from viewport and/or field 1`] = `
<div
  className="makeStyles-root-1 makeStyles-root-4"
>
  <div
    className="makeStyles-column-2"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-root-13 makeStyles-root-18 makeStyles-closed-16"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
         
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`GameField component should take correct dimensions from viewport and/or field 2`] = `
<div
  className="makeStyles-root-1 makeStyles-root-19"
>
  <div
    className="makeStyles-column-2"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-root-13 makeStyles-root-20 makeStyles-closed-16"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
         
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`GameField component should take correct dimensions from viewport and/or field 3`] = `
<div
  className="makeStyles-root-1 makeStyles-root-21"
>
  <div
    className="makeStyles-column-2"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-root-13 makeStyles-root-22 makeStyles-closed-16"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
         
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`GameField component should take correct dimensions from viewport and/or field 4`] = `
<div
  className="makeStyles-root-1 makeStyles-root-23"
>
  <div
    className="makeStyles-column-2"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-root-13 makeStyles-root-24 makeStyles-closed-16"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
         
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`GameField component should take correct dimensions from viewport and/or field 5`] = `
<div
  className="makeStyles-root-1 makeStyles-root-25"
>
  <div
    className="makeStyles-column-2"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-root-13 makeStyles-root-26 makeStyles-closed-16"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
         
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`GameField component should take correct dimensions from viewport and/or field 6`] = `
<div
  className="makeStyles-root-1 makeStyles-root-27"
>
  <div
    className="makeStyles-column-2"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-root-13 makeStyles-root-28 makeStyles-closed-16"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
         
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`GameField component should take correct dimensions from viewport and/or field 7`] = `
<div
  className="makeStyles-root-1 makeStyles-root-29"
>
  <div
    className="makeStyles-column-2"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-root-13 makeStyles-root-30 makeStyles-closed-16"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
         
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`GameField component should take correct dimensions from viewport and/or field 8`] = `
<div
  className="makeStyles-root-1 makeStyles-root-31"
>
  <div
    className="makeStyles-column-2"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-root-13 makeStyles-root-32 makeStyles-closed-16"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
         
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`GameField component should take correct dimensions from viewport and/or field 9`] = `
<div
  className="makeStyles-root-1 makeStyles-root-33"
>
  <div
    className="makeStyles-column-2"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-root-13 makeStyles-root-34 makeStyles-closed-16"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
         
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
